yjit/src/core.rs

```diff
@@ -3455,6 +3472,44 @@ fn branch_stub_hit_body(branch_ptr: *const c_void, target_idx: u32, ec: EcPtr) -
             // Branch shape should reflect layout
             assert!(!(branch.gen_fn.get_shape() == target_branch_shape && new_block.start_addr != branch.end_addr.get()));
 
+            // If we've hit a deferred branch, and the housing block is solely the branch,
+            // rewire incoming branches to the new block and delete the housing block.
+            {
+                // This &Block should be unique, relying on the VM lock
+                let housing_block: &Block = unsafe { branch.block.get().as_ref() };
+                if target_ctx.is_deferred() &&
+                    target_blockid == housing_block.get_blockid() &&
+                    housing_block.outgoing.len() == 1
+                {
+                    // Divert incoming branches of housing_block to the new block
+                    housing_block.incoming.for_each(|incoming| {
+                        let incoming = unsafe { incoming.as_ref() };
+                        for target in 0..incoming.targets.len() {
+                            // SAFETY: No cell mutation; copying out a BlockRef.
+                            if Some(BlockRef::from(housing_block)) == unsafe {
+                                        incoming.targets[target]
+                                            .ref_unchecked()
+                                            .as_ref()
+                                            .and_then(|target| target.get_block())
+                                    } {
+                                incoming.targets[target].set(Some(Box::new(BranchTarget::Block(new_block.into()))));
+                            }
+                        }
+                        new_block.push_incoming(incoming.into());
+                    });
+
+                    // Transplant the branch we've just hit to the new block
+                    mem::drop(housing_block.outgoing.0.take());
+                    new_block.outgoing.push(branch_ref);
+                    let housing_block: BlockRef = branch.block.replace(new_block.into());
+                    // Free the old housing block. There should now be no live &Block.
+                    remove_block_version(&housing_block);
+                    unsafe { free_block(housing_block, false) };
+
+                    incr_counter!(deleted_defer_block_count);
+                }
+            }
```

Comment:
`branch_stub_hit_body` is already a huge function. Do you think this logic could be moved into its own function to make the code more readable?